fixed A13 and A15 tests

makslion opened the pull request:
# Description
Part of the 1488 JIRA
Fixed A13 and A15 tests to get replicas from CRs and compare them against a number of pods running on a cluster for products, configured by Quota 

## Verification steps
1. ensure you have access to a cluster with RHOAM installed from a feature branch: log in to the cluster.
2. run the following commands in your terminal:
```
INSTALLATION_TYPE=managed-api TEST=A13 make test/e2e/single
INSTALLATION_TYPE=managed-api TEST=A14 make test/e2e/single
INSTALLATION_TYPE=managed-api TEST=A15 make test/e2e/single
```


## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update <!-- Update JIRA with Affects -> Documentation -->
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

matskiv commented on test/common/deployment_types.go:
```diff
@@ -2,6 +2,13 @@ package common
+	"github.com/integr8ly/integreatly-operator/pkg/resources/quota"
+	"github.com/keycloak/keycloak-operator/pkg/apis/keycloak/v1alpha1"
+	"golang.org/x/net/context"
+	v12 "k8s.io/api/apps/v1"
+	v13 "k8s.io/api/core/v1"
```
Can you rename these imports please?
`v13` -> `corev1`
and `v12` to something like `k8sappsv1` ?

KevFan commented on test/common/deployment_types.go:
```diff
@@ -183,6 +189,28 @@ func getDeploymentConfiguration(deploymentName string, inst *integreatlyv1alpha1
+				t.Fatalf("Error obtaining ratelimit CR: %v", err)
+			}
+		}
+		deployment["marin3rDeployment"].Products[1].ExpectedReplicas = *ratelimitCR.Spec.Replicas
```
```suggestion
		deployment["marin3rDeployment"] = Namespace{
			Name: Marin3rProductNamespace,
			Products: []Product{
				{Name: "prom-statsd-exporter", ExpectedReplicas: 1},
				{Name: "ratelimit", ExpectedReplicas: *ratelimitCR.Spec.Replicas},
			},
		}
```
What do you think on just adding the `deployment["marin3rDeployment"]` block from https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L149-L155 to  here instead of just changing the replica value ?  I feel it would be slighly safer on the off chance someone in the future adds to this and messes up the array order causing the wrong struct replica value to be reassigned. Also this will make it a single place where it is added and not changed somewhere else in the code :thinking:

makslion replied:
Yes, that's definitely the change to be made. Thanks

matskiv commented on test/common/deployment_types.go:
```diff
@@ -183,6 +189,28 @@ func getDeploymentConfiguration(deploymentName string, inst *integreatlyv1alpha1
+		}
+
+		key, err := k8sclient.ObjectKeyFromObject(ratelimitCR)
+		if err != nil {
+			t.Fatalf("Error getting key from ratelimit Deployment: %v", err)
+		}
```
If I am looking at the correct source code for `ObjectKeyFromObject()`,
then we don't need any error checking as it never returns errors:
https://github.com/kubernetes-sigs/controller-runtime/blob/b2c90ab82af89fb84120108af14f4ade9df5d787/pkg/client/interfaces.go#L33-L36

makslion replied:
here is an implementation https://github.com/makslion/integreatly-operator/blob/MGDAPI-1488-A13/vendor/sigs.k8s.io/controller-runtime/pkg/client/interfaces.go#L33-L39

matskiv replied:
Oh, I was looking at newer version of controller-runtime...
Sorry for spam.. 😅

KevFan commented on test/common/deployment_types.go:
```diff
@@ -376,6 +478,25 @@ func TestStatefulSetsExpectedReplicas(t TestingTB, ctx *TestingContext) {
+		keycloakCR := &v1alpha1.Keycloak{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      quota.KeycloakName,
+				Namespace: NamespacePrefix + "user-sso",
```
```suggestion
				Namespace: RHSSOUserProductNamespace,
```
Assuming rebase with the latest changes from feature branch

makslion replied:
Missed those :see_no_evil:

KevFan commented on test/common/deployment_types.go:
```diff
@@ -416,6 +537,30 @@ func TestStatefulSetsExpectedReplicas(t TestingTB, ctx *TestingContext) {
+				pods := &v13.PodList{}
+				selector, _ := labels.Parse("component=keycloak")
+				keycloakListOpts := []k8sclient.ListOption{
+					k8sclient.InNamespace(NamespacePrefix + "user-sso"),
```
```suggestion
					k8sclient.InNamespace(RHSSOUserProductNamespace),
```
Assuming rebase with the latest changes from feature branch

KevFan reviewed (commented):
Overall looks good :+1: Left some comments / questions

KevFan commented on test/common/deployment_types.go:
```diff
@@ -266,39 +294,113 @@ func TestDeploymentExpectedReplicas(t TestingTB, ctx *TestingContext) {
+						k8sclient.InNamespace(ThreeScaleProductNamespace),
+						k8sclient.MatchingLabelsSelector{
+							Selector: selector,
+						},
+					}
+					err = ctx.Client.List(context.TODO(), pods, listOpts...)
+					if err != nil {
+						t.Fatalf("failed to get backend listener pods for 3scale: %v", err)
+					}
+					checkPods(t, pods, product, namespace, deployment)
+
+				} else if product.Name == "backend-listener" {
+					pods := &v13.PodList{}
+					selector, _ := labels.Parse("deploymentConfig=backend-listener")
+					listOpts := []k8sclient.ListOption{
+						k8sclient.InNamespace(ThreeScaleProductNamespace),
+						k8sclient.MatchingLabelsSelector{
+							Selector: selector,
+						},
+					}
+					err = ctx.Client.List(context.TODO(), pods, listOpts...)
+					if err != nil {
+						t.Fatalf("failed to get backend listener pods for 3scale: %v", err)
+					}
+					checkPods(t, pods, product, namespace, deployment)
+
+				} else if product.Name == "backend-worker" {
+					pods := &v13.PodList{}
+					selector, _ := labels.Parse("deploymentconfig=backend-worker")
+					listOpts := []k8sclient.ListOption{
+						k8sclient.InNamespace(ThreeScaleProductNamespace),
+						k8sclient.MatchingLabelsSelector{
+							Selector: selector,
+						},
+					}
+					err = ctx.Client.List(context.TODO(), pods, listOpts...)
+					if err != nil {
+						t.Fatalf("failed to get backend listener pods for 3scale: %v", err)
+					}
+					checkPods(t, pods, product, namespace, deployment)
```
Bit confused on why there's these blocks for checking 3scale deployment configs here ? :thinking: 
This specific test `TestDeploymentExpectedReplicas` is for `deployments` and I don't think it would ever get 3scale deployment configs in this test 
* https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L261
* https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L385-L405
* https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L24-L45

Should also fail here since deployment resource would not exist for 3scale deployment configs :thinking: 
* https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L280-L285

Were you meaning to put these code blocks into the `TestDeploymentConfigExpectedReplicas` test ? :thinking: 
* https://github.com/integr8ly/integreatly-operator/blob/3ae07fc47fd73ee86a7e18c86054b8d2f40c709b/test/common/deployment_types.go#L407-L451

makslion replied:
Now I'm wondering how I had those passing. Well spotted. Moved 3scale checks to the `TestDeploymentConfigExpectedReplicas`.

makslion commented:
rebased against feature branch to include latest changes, implemented suggestion from reviewers and removed few redundant declarations.

makslion commented on test/common/deployment_types.go:
```diff
@@ -266,39 +300,80 @@ func TestDeploymentExpectedReplicas(t TestingTB, ctx *TestingContext) {
-					product.ExpectedReplicas,
-				)
-				continue
+			if rhmi.Spec.Type == string(integreatlyv1alpha1.InstallationTypeManagedApi) {
+				if product.Name == "ratelimit" {
+					pods := &corev1.PodList{}
+					selector, _ := labels.Parse("app=ratelimit")
+					listOpts := []k8sclient.ListOption{
+						k8sclient.InNamespace(Marin3rProductNamespace),
+						k8sclient.MatchingLabelsSelector{
+							Selector: selector,
+						},
+					}
+					err = ctx.Client.List(context.TODO(), pods, listOpts...)
+					if err != nil {
+						t.Fatalf("failed to get pods for Ratelimit: %v", err)
+					}
+					checkDeploymentPods(t, pods, product, namespace, deployment)
+
+				}
+			} else {
+				// Verify that the expected replicas are also available, means they are up and running and consumable by users
+				if deployment.Status.AvailableReplicas < product.ExpectedReplicas {
+					t.Errorf("Deployment %s in namespace %s doesn't match the number of expected available replicas. Available Replicas: %v / Expected Replicas: %v",
+						product.Name,
+						namespace.Name,
+						deployment.Status.AvailableReplicas,
+						product.ExpectedReplicas,
+					)
+					continue
+				}
 			}
```
and in this case those two `if` statements should be combined into one, yes :thinking:

KevFan commented on test/common/deployment_types.go:
```diff
@@ -93,61 +99,61 @@ func getDeploymentConfiguration(deploymentName string, inst *integreatlyv1alpha1
 			},
 		},
-		"marin3rDeployment": Namespace{
+		"marin3rDeployment": {
 			Name: Marin3rProductNamespace,
 			Products: []Product{
 				{Name: "prom-statsd-exporter", ExpectedReplicas: 1},
 				{Name: "ratelimit", ExpectedReplicas: 3},
 			},
```
We could remove this since it's later assigned and only applies to managed api

makslion replied:
Removed. You are absolutely right.

KevFan commented on test/common/deployment_types.go:
```diff
@@ -334,8 +407,53 @@ func TestDeploymentConfigExpectedReplicas(t TestingTB, ctx *TestingContext) {
 			}
-
-			if deploymentConfig.Status.AvailableReplicas < product.ExpectedReplicas {
+			if rhmi.Spec.Type == string(integreatlyv1alpha1.InstallationTypeManagedApi) {
```
Do we need this `if` block in general ? Seems we are just checking number of pods running vs expected number ? I don't there'll be much harm in having this run in RHMI e2e also unless I'm missing something ? :thinking:

makslion replied:
Not sure about that one. The reason why we decided to bother ourselves with checking values in pods in the first place was an introduction of Quota. Idea was to check not only that we set values in CRs but also that we are setting values that could be accepted by various products and could be propagated down to pods. 
For RHMI nothing gets changed so the default CR check should be sufficient. In theory, nothing holds us from checking CPU and memory in pods for RHMI (the code to do so is already here), but I think that if we decide to refactor RHMI-related things they, probably, should be addressed in a separate ticket :thinking: 
What do you think?

KevFan replied:
Ah okay, just wondering overall. I'm good keeping it as it if we want to isolate this to just RHOAM :+1:

KevFan commented on test/common/deployment_types.go:
```diff
@@ -334,8 +407,53 @@ func TestDeploymentConfigExpectedReplicas(t TestingTB, ctx *TestingContext) {
-			if deploymentConfig.Status.AvailableReplicas < product.ExpectedReplicas {
+			if rhmi.Spec.Type == string(integreatlyv1alpha1.InstallationTypeManagedApi) {
+				if product.Name == "apicast-production" {
+					pods := &corev1.PodList{}
+					selector, _ := labels.Parse("deploymentconfig=apicast-production")
+					listOpts := []k8sclient.ListOption{
+						k8sclient.InNamespace(ThreeScaleProductNamespace),
+						k8sclient.MatchingLabelsSelector{
+							Selector: selector,
+						},
+					}
+					err = ctx.Client.List(context.TODO(), pods, listOpts...)
+					if err != nil {
+						t.Fatalf("failed to get backend listener pods for 3scale: %v", err)
+					}
+					checkDeploymentConfigPods(t, pods, product, namespace, deploymentConfig)
+
```
What do you think on having an helper function for getting returning `listOpts` ? Something like
```
func getListOptsForSelectorInNamespace(label, namespace string) []k8sclient.ListOption {
	selector, _ := labels.Parse(label)
	listOpts := []k8sclient.ListOption{
		k8sclient.InNamespace(namespace),
		k8sclient.MatchingLabelsSelector{
			Selector: selector,
		},
	}
	
	return listOpts
}
```
Then the `if` blocks would be smaller in size :thinking: 
```suggestion
					pods := &corev1.PodList{}
					listOpts := getListOptsForSelectorInNamespace("deploymentconfig=apicast-production", ThreeScaleProductNamespace)
					err = ctx.Client.List(context.TODO(), pods, listOpts...)
					if err != nil {
						t.Fatalf("failed to get backend listener pods for 3scale: %v", err)
					}
					checkDeploymentConfigPods(t, pods, product, namespace, deploymentConfig)

```

makslion replied:
yes, it would make it prettier. Let me try to make it up.

KevFan reviewed (commented):
Nice :+1: Just a few final questions / comments :technologist:

makslion commented:
Implemented more suggestions. Also modified description to check A14 as it is also got affected by this PR. 
Introduced `GetListOptions(...)` function to obtain list options for getting pods. Made it public as it will be used during the testing week to refactor existing code in the `test` package that as well utilizes list options for pods retrieval

makslion commented:
/test
hey, bot, would you do the thing? :thinking:

makslion commented:
/test all

trepel commented:
@makslion I have a cluster installed from https://github.com/integr8ly/integreatly-operator/pull/1739 since I was reviewing that one. I ran the A13-15 tests against that cluster from this PR locally and all passed.

I am adding lgtm but I'd be good if @matskiv and @KevFan have another look if their comments were addressed properly.

/lgtm

KevFan commented:
Code looks good to me :+1: 
/lgtm
